File: Sources/PetCompanion/Pet/PetMovementController.swift
import AppKit
import SpriteKit

final class PetMovementController {
    private(set) var direction: CGFloat = 1
    private var dockEdge: NSRectEdge = .minY
    private var dockBaseline: CGFloat = 13
    private var homePosition: CGPoint?

    func configure(for dockEdge: NSRectEdge, dockBaseline: CGFloat) {
        self.dockEdge = dockEdge
        self.dockBaseline = dockBaseline
    }

    func startingPosition(in size: CGSize, petSize: CGSize) -> CGPoint {
        let position: CGPoint
        switch dockEdge {
        case .minY:
            position = CGPoint(x: max(petSize.width, size.width * 0.5), y: dockBaseline)
        case .minX:
            position = CGPoint(x: dockBaseline, y: max(petSize.height, size.height * 0.5))
        case .maxX:
            position = CGPoint(x: dockBaseline, y: max(petSize.height, size.height * 0.5))
        default:
            position = CGPoint(x: max(petSize.width, size.width * 0.5), y: dockBaseline)
        }
        homePosition = position
        return position
    }

    func dockAlignedPosition(currentPosition: CGPoint, sceneSize: CGSize, petSize: CGSize) -> CGPoint {
        var next = currentPosition

        switch dockEdge {
        case .minY:
            let minX = petSize.width * 0.55
            let maxX = max(minX, sceneSize.width - petSize.width * 0.55)
            next.x = min(max(next.x, minX), maxX)
            next.y = dockBaseline
        case .minX:
            let minY = petSize.height * 0.55
            let maxY = max(minY, sceneSize.height - petSize.height * 0.55)
            next.x = dockBaseline
            next.y = min(max(next.y, minY), maxY)
        case .maxX:
            let minY = petSize.height * 0.55
            let maxY = max(minY, sceneSize.height - petSize.height * 0.55)
            next.x = dockBaseline
            next.y = min(max(next.y, minY), maxY)
        default:
            break
        }

        homePosition = clampedHome(from: homePosition ?? next, sceneSize: sceneSize, petSize: petSize)
        return next
    }

    func setHomePosition(_ position: CGPoint, sceneSize: CGSize, petSize: CGSize) -> CGPoint {
        let clamped = clampedFreePosition(position, sceneSize: sceneSize, petSize: petSize)
        homePosition = clamped
        return clamped
    }

    func update(position: CGPoint, sceneSize: CGSize, petSize: CGSize, speed: CGFloat, deltaTime: TimeInterval) -> CGPoint {
        var next = position
        let distance = speed * CGFloat(deltaTime)

        switch dockEdge {
        case .minY:
            next.x += direction * distance
            let bounds = walkingBounds(sceneSize: sceneSize, petSize: petSize)
            let minX = bounds.lowerBound
            let maxX = bounds.upperBound
            if next.x <= minX {
                next.x = minX
                direction = 1
            } else if next.x >= maxX {
                next.x = maxX
                direction = -1
            }
        case .minX, .maxX:
            next.y += direction * distance
            let bounds = walkingBounds(sceneSize: sceneSize, petSize: petSize)
            let minY = bounds.lowerBound
            let maxY = bounds.upperBound
            if next.y <= minY {
                next.y = minY
                direction = 1
            } else if next.y >= maxY {
                next.y = maxY
                direction = -1
            }
        default:
            break
        }

        return next
    }

    private func walkingBounds(sceneSize: CGSize, petSize: CGSize) -> ClosedRange<CGFloat> {
        switch dockEdge {
        case .minY:
            let fullMin = petSize.width * 0.55
            let fullMax = max(fullMin, sceneSize.width - petSize.width * 0.55)
            let center = clampedHome(from: homePosition ?? CGPoint(x: sceneSize.width * 0.5, y: dockBaseline), sceneSize: sceneSize, petSize: petSize).x
            let halfRange = sceneSize.width / 6
            return max(fullMin, center - halfRange)...min(fullMax, center + halfRange)
        case .minX, .maxX:
            let fullMin = petSize.height * 0.55
            let fullMax = max(fullMin, sceneSize.height - petSize.height * 0.55)
            let center = clampedHome(from: homePosition ?? CGPoint(x: dockBaseline, y: sceneSize.height * 0.5), sceneSize: sceneSize, petSize: petSize).y
            let halfRange = sceneSize.height / 6
            return max(fullMin, center - halfRange)...min(fullMax, center + halfRange)
        default:
            return 0...0
        }
    }

    private func clampedHome(from position: CGPoint, sceneSize: CGSize, petSize: CGSize) -> CGPoint {
        var next = position

        switch dockEdge {
        case .minY:
            let minX = petSize.width * 0.55
            let maxX = max(minX, sceneSize.width - petSize.width * 0.55)
            next.x = min(max(next.x, minX), maxX)
            next.y = dockBaseline
        case .minX:
            let minY = petSize.height * 0.55
            let maxY = max(minY, sceneSize.height - petSize.height * 0.55)
            next.x = dockBaseline
            next.y = min(max(next.y, minY), maxY)
        case .maxX:
            let minY = petSize.height * 0.55
            let maxY = max(minY, sceneSize.height - petSize.height * 0.55)
            next.x = dockBaseline
            next.y = min(max(next.y, minY), maxY)
        default:
            break
        }

        return next
    }

    private func clampedFreePosition(_ position: CGPoint, sceneSize: CGSize, petSize: CGSize) -> CGPoint {
        let minX = petSize.width * 0.55
        let maxX = max(minX, sceneSize.width - petSize.width * 0.55)
        let minY = petSize.height * 0.55
        let maxY = max(minY, sceneSize.height - petSize.height * 0.55)

        return CGPoint(
            x: min(max(position.x, minX), maxX),
            y: min(max(position.y, minY), maxY)
        )
    }
}

File: Windows/PetCompanion.Windows/Pet/PetMovementController.cs
using System;
using System.Windows;
using PetCompanion.Windows.Platform;

namespace PetCompanion.Windows.Pet;

public sealed class PetMovementController
{
    private TaskbarEdge edge = TaskbarEdge.Bottom;
    private double baseline = 13;
    private Point? homePosition;

    public double Direction { get; private set; } = 1;

    public void Configure(TaskbarEdge edge, double baseline)
    {
        this.edge = edge;
        this.baseline = baseline;
    }

    public Point StartingPosition(Size sceneSize, Size petSize)
    {
        var position = edge switch
        {
            TaskbarEdge.Bottom => new Point(Math.Max(petSize.Width, sceneSize.Width * 0.5), baseline),
            TaskbarEdge.Top => new Point(Math.Max(petSize.Width, sceneSize.Width * 0.5), baseline),
            TaskbarEdge.Left => new Point(baseline, Math.Max(petSize.Height, sceneSize.Height * 0.5)),
            TaskbarEdge.Right => new Point(baseline, Math.Max(petSize.Height, sceneSize.Height * 0.5)),
            _ => new Point(Math.Max(petSize.Width, sceneSize.Width * 0.5), baseline)
        };
        homePosition = position;
        return position;
    }

    public Point TaskbarAlignedPosition(Point currentPosition, Size sceneSize, Size petSize)
    {
        var next = currentPosition;
        switch (edge)
        {
            case TaskbarEdge.Bottom:
            case TaskbarEdge.Top:
                next.X = Clamp(next.X, petSize.Width * 0.55, sceneSize.Width - petSize.Width * 0.55);
                next.Y = baseline;
                break;
            case TaskbarEdge.Left:
            case TaskbarEdge.Right:
                next.X = baseline;
                next.Y = Clamp(next.Y, petSize.Height * 0.55, sceneSize.Height - petSize.Height * 0.55);
                break;
        }
        homePosition = ClampHome(homePosition ?? next, sceneSize, petSize);
        return next;
    }

    public Point SetHomePosition(Point position, Size sceneSize, Size petSize)
    {
        var clamped = ClampFreePosition(position, sceneSize, petSize);
        homePosition = clamped;
        return clamped;
    }

    public Point Update(Point position, Size sceneSize, Size petSize, double speed, double deltaTime)
    {
        var next = position;
        var distance = speed * deltaTime;
        var bounds = WalkingBounds(sceneSize, petSize);

        if (edge is TaskbarEdge.Bottom or TaskbarEdge.Top)
        {
            next.X += Direction * distance;
            if (next.X <= bounds.Min)
            {
                next.X = bounds.Min;
                Direction = 1;
            }
            else if (next.X >= bounds.Max)
            {
                next.X = bounds.Max;
                Direction = -1;
            }
        }
        else
        {
            next.Y += Direction * distance;
            if (next.Y <= bounds.Min)
            {
                next.Y = bounds.Min;
                Direction = 1;
            }
            else if (next.Y >= bounds.Max)
            {
                next.Y = bounds.Max;
                Direction = -1;
            }
        }

        return next;
    }

    private (double Min, double Max) WalkingBounds(Size sceneSize, Size petSize)
    {
        if (edge is TaskbarEdge.Bottom or TaskbarEdge.Top)
        {
            var fullMin = petSize.Width * 0.55;
            var fullMax = Math.Max(fullMin, sceneSize.Width - petSize.Width * 0.55);
            var center = ClampHome(homePosition ?? new Point(sceneSize.Width * 0.5, baseline), sceneSize, petSize).X;
            var halfRange = sceneSize.Width / 6;
            return (Math.Max(fullMin, center - halfRange), Math.Min(fullMax, center + halfRange));
        }
        else
        {
            var fullMin = petSize.Height * 0.55;
            var fullMax = Math.Max(fullMin, sceneSize.Height - petSize.Height * 0.55);
            var center = ClampHome(homePosition ?? new Point(baseline, sceneSize.Height * 0.5), sceneSize, petSize).Y;
            var halfRange = sceneSize.Height / 6;
            return (Math.Max(fullMin, center - halfRange), Math.Min(fullMax, center + halfRange));
        }
    }

    private Point ClampHome(Point position, Size sceneSize, Size petSize)
    {
        return edge switch
        {
            TaskbarEdge.Bottom or TaskbarEdge.Top => new Point(
                Clamp(position.X, petSize.Width * 0.55, sceneSize.Width - petSize.Width * 0.55),
                baseline),
            TaskbarEdge.Left or TaskbarEdge.Right => new Point(
                baseline,
                Clamp(position.Y, petSize.Height * 0.55, sceneSize.Height - petSize.Height * 0.55)),
            _ => position
        };
    }

    private static Point ClampFreePosition(Point position, Size sceneSize, Size petSize)
    {
        return new Point(
            Clamp(position.X, petSize.Width * 0.55, sceneSize.Width - petSize.Width * 0.55),
            Clamp(position.Y, petSize.Height * 0.55, sceneSize.Height - petSize.Height * 0.55));
    }

    private static double Clamp(double value, double min, double max) => Math.Min(Math.Max(value, min), max);
}
